library/src/main/java/SpecTreesUnionizer.java

```diff
@@ -0,0 +1,256 @@
+
+    conflicts.removeIf(conflict -> defaultKeypaths.contains(conflict.getKeypath()));
+  }
+
+  /**
+   * Performs a union on {@code defaults} and {@code map2} and returns the result. Conflicts are
+   * resolved by using the {@code defaults} map as priority.
+   *
+   * @param defaults the map which {@code map2} is merged into. {@code defaults} will contain the
+   *     result of the union and will take priority over {@code map2} in the case of conflicts.
+   * @param map2 the map to merge into {@code defaults}
+   * @return the result of a union on {@code defaults} and {@code map2}, where conflicts are
+   *     resolved by using the {@code defaults} map as priority.
+   */
+  static LinkedHashMap<String, Object> applyOverlay(
+      LinkedHashMap<String, Object> defaults, LinkedHashMap<String, Object> map2)
+      throws UnexpectedTypeException {
+    return union(
+        defaults,
+        map2,
+        true,
+        new Stack<String>(),
+        new ArrayList<Conflict>(),
+        new HashMap<String, Object>());
+  }
+
+  /**
+   * Union function with all possible options. Other functions provide a nicer interface for
+   * different use cases of union, and ultimately call this function.
+   *
+   * @param map1 map 1/2 to merge. This map is special because it can take priority in the union
+   *     based on {@code map1IsDefault} flag
+   * @param map2 map 2/2 to merge
+   * @param map1IsDefault if true, map1 will take priority over map2 in case of different leaf
+   *     values, and no conflict will be reported
+   * @param keypath the key path which the leaf nodes belong to in the current iteration of the
+   *     recursive frame
+   * @param conflicts appended to if there is an unresolvable conflict
+   * @param conflictResolutions a map which can provide conflict resolutions based on keypaths
+   * @return the result of union on {@code map1} and {@code map2}
+   */
+  private static LinkedHashMap<String, Object> union(
+      LinkedHashMap<String, Object> map1,
+      LinkedHashMap<String, Object> map2,
+      boolean map1IsDefault,
+      Stack<String> keypath,
+      ArrayList<Conflict> conflicts,
+      HashMap<String, Object> conflictResolutions)
+      throws UnexpectedTypeException {
+
+    for (Map.Entry<String, Object> entry : map2.entrySet()) {
+      String key = entry.getKey();
+      Object value2 = entry.getValue();
+
+      keypath.push(key);
+
+      if (map1.containsKey(entry.getKey())) {
+        Object value1 = map1.get(key);
+
+        if (!value1.equals(value2)) {
+          if (TypeChecker.isObjectMap(value1) && TypeChecker.isObjectMap((value2))) {
+            // We have two maps which contain the same key, add the key and process maps further.
+            processUnequalSubtrees(
+                map1, map1IsDefault, keypath, conflicts, conflictResolutions, key, value2, value1);
+          } else if (TypeChecker.isObjectList(value1)
```

Comment:
Nesting is getting a little deep here. Consider reworking into smaller methods.

Reply:
addressed!